File: app/src/main/java/com/appdocker/vertx/ServiceDiscoveryVerticle.java
package com.appdocker.vertx;


import io.vertx.core.AbstractVerticle;
import io.vertx.core.Future;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import io.vertx.core.logging.Logger;
import io.vertx.core.logging.LoggerFactory;
import io.vertx.servicediscovery.Record;
import io.vertx.servicediscovery.ServiceDiscovery;
import io.vertx.servicediscovery.types.EventBusService;

final class ServiceDiscoveryVerticle extends AbstractVerticle {

    private static final Logger logger = LoggerFactory.getLogger(ServiceDiscoveryVerticle.class);

    private ServiceDiscovery discovery = null;


    @Override
    public void start(Future<Void> startFuture) throws Exception {



        discovery = ServiceDiscovery.create(vertx);

        createExport(config().getJsonArray("export"));

        createImport(config().getJsonArray("import"));

        super.start(startFuture);
    }

    private void createExport(JsonArray services) {

        if (services == null) return;

        for(Object object : services.getList()) {

            JsonObject jsonObject = (JsonObject)object;

            Record record = EventBusService.createRecord(
                    jsonObject.getString("name"),
                    jsonObject.getString("address"),
                    jsonObject.getString("service"),
                    jsonObject.getJsonObject("metadata"));

            discovery.publish(record,ar -> {
                if (!ar.succeeded()) {
                    logger.error("publish service error .",ar.cause());
                }
            });
        }
    }

    private void createImport(JsonArray services) {

        if (services == null) return;

        for(Object object : services.getList()) {

            discovery.getRecord((JsonObject) object, ar -> {
                if (ar.succeeded()) {
                    if (ar.result() != null) {
                        logger.info("service({0}) status {1} type {2}",ar.result().getName(),ar.result().getStatus(),ar.result().getType());
                    }
                } else {
                    logger.info("getRecord error .",ar.cause());
                }
            });
        }
    }
}

File: appdocker/src/main/kotlin/com/appdocker/ServiceDiscoveryVerticle.kt
package com.appdocker

import io.vertx.core.AbstractVerticle
import io.vertx.core.json.JsonObject
import io.vertx.core.logging.LoggerFactory
import io.vertx.servicediscovery.Record
import io.vertx.servicediscovery.ServiceDiscovery
import io.vertx.servicediscovery.types.EventBusService
import io.vertx.serviceproxy.ProxyHelper
import java.util.concurrent.ConcurrentHashMap


class ServiceDiscoveryVerticle:AbstractVerticle() {

    var discovery:ServiceDiscovery? = null

    val published = ConcurrentHashMap<JsonObject,Record>()

    val logger = LoggerFactory.getLogger(ServiceDiscoveryVerticle::class.java)!!

    override fun stop() {
        super.stop()
    }

    override fun start() {

        discovery = ServiceDiscovery.create(vertx)

        // start service discovery
        exportService()
    }

    private fun exportService() {

        val exports = config().getJsonArray("export") ?: return

        exports
                .filterIsInstance<JsonObject>()
                .forEach { export ->

                    val record = EventBusService.createRecord(
                            export.getString("name"),
                            export.getString("address"),
                            export.getString("service"),
                            export.getJsonObject("metadata"))

                    discovery!!.publish(record,{
                        ar ->

                        if (ar.succeeded()) {

                            published[export] = ar.result()

                            //                    // reflect create service implement
                            val implClassName = export.getString("impl")

                            if (implClassName == null) {
                                logger.error("export service must indicate impl class\n\t",export)
                                System.exit(1)
                            }

                            val implClass = Class.forName(implClassName)

                            val obj = implClass.newInstance()

                            @Suppress("UNCHECKED_CAST")
                            ProxyHelper.registerService(Class.forName(export.getString("service")) as Class<Any>,vertx,obj,export.getString("address"))

                            logger.info("create event bus service -- success :\n\t{}",export)
                        } else {
                            logger.error("create event bus service -- failed :\n\t{}",export,ar.cause())
                        }
                    })
                }
    }
}